Get UnsolvedSymbolException when trying to resolve type

Hi people,

I am facing an issue at this moment,  a `UnsolvedSymbolException` is thrown when resolving ```FieldAccessExpr``` object.

Basically, I want to resolve the type of an object which field is accessed. You could reproduce the error by the following steps.

- Create a java Class `ClassA`
```
public class ClassA {

    public ClassB obj;
}
```

- Create a second java class `ClassB`

```
public class ClassB {

    public Long value;
}
```

- Create a third class `TestClass` which is the java file I would like to parse.
```
public class TestClass {

    public Long getValue() {

        Optional<ClassA> classA = Optional.of(new ClassA());

        return classA.map(a -> a.obj)
                .map(b -> b.value)
                .orElse(null);
    }
}
```
- Create a class to parse the  `TestClass.java` and resolve type of objects which filed is accessed. In this example, my goal is to resolve what are the types of objects which field are accessed in the `getValue` method. I hide the real file path name in the code snippet below. 

```
public class Inspect {

    public static void main(String[] args) throws FileNotFoundException {
        JavaSymbolSolver javaSymbolSolver = createSymbolSolver();
        StaticJavaParser.getConfiguration().setSymbolResolver(javaSymbolSolver);
        String path = "PATH_TO_TestClass.java";
        CompilationUnit cu = StaticJavaParser.parse(new File(path));
        cu.findAll(FieldAccessExpr.class).forEach(
                FieldAccessExpr::resolve
        );
    }

    private static JavaSymbolSolver createSymbolSolver() {
        TypeSolver reflectionTypeSolver = new ReflectionTypeSolver();
        JavaParserTypeSolver javaParserTypeSolver = new JavaParserTypeSolver(new File("PATH_TO_ROO_PACKAGE"));
        CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver(reflectionTypeSolver, javaParserTypeSolver);
        return new JavaSymbolSolver(combinedTypeSolver);
    }
}
```

The exception I got when executing `Inspect` is 
```
Caused by: UnsolvedSymbolException{context='null', name='obj', cause='null'}
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:237)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:64)
	at com.github.javaparser.ast.expr.FieldAccessExpr.accept(FieldAccessExpr.java:90)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:526)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:358)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:340)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:475)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:64)
	at com.github.javaparser.ast.expr.LambdaExpr.accept(LambdaExpr.java:162)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:526)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:358)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:367)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:340)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AbstractJavaParserContext.findTypeDeclarations(AbstractJavaParserContext.java:172)
	... 26 more
```

I also noticed that types can be resolved if I change the `getValue` method of `TestClass` as below 

```
import java.util.Optional;

public class TestClass {

    public Long getValue() {

        Optional<ClassA> classA = Optional.of(new ClassA());

        final Optional<ClassB> classB = classA.map(a -> a.obj);
        return classB
                .map(b -> b.value)
                .orElse(null);
    }
}
```

As you can see that the `map` methods are not chained , instead, the return of first `map` method is assigned explicitly to a local variable. To me, it seems the types can NOT be resolved if there is a chain of method invocations and methods have lambda function as input parameters. 


I don't know if I misused the library or I am missing some configurations. Could anyone please provide some suggestions? I really appreciate it!

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/FieldAccessContext.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/FieldAccessContext.java
@@ -21,6 +21,12 @@
 
 package com.github.javaparser.symbolsolver.javaparsermodel.contexts;
 
+import static com.github.javaparser.symbolsolver.javaparser.Navigator.demandParentNode;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
 import com.github.javaparser.ast.expr.Expression;
 import com.github.javaparser.ast.expr.FieldAccessExpr;
 import com.github.javaparser.ast.expr.ThisExpr;
@@ -32,6 +38,7 @@ import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclar
 import com.github.javaparser.resolution.declarations.ResolvedTypeDeclaration;
 import com.github.javaparser.resolution.declarations.ResolvedValueDeclaration;
 import com.github.javaparser.resolution.types.ResolvedPrimitiveType;
+import com.github.javaparser.resolution.types.ResolvedReferenceType;
 import com.github.javaparser.resolution.types.ResolvedType;
 import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
 import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFactory;
@@ -39,12 +46,6 @@ import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
 import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
 import com.github.javaparser.symbolsolver.model.resolution.Value;
 import com.github.javaparser.symbolsolver.resolution.SymbolSolver;
-
-import java.util.Collection;
-import java.util.List;
-import java.util.Optional;
-
-import static com.github.javaparser.symbolsolver.javaparser.Navigator.demandParentNode;
 
 /**
  * @author Federico Tomassetti
@@ -92,18 +93,9 @@ public class FieldAccessContext extends AbstractJavaParserContext<FieldAccessExp
                 return Optional.of(new Value(ResolvedPrimitiveType.INT, ARRAY_LENGTH_FIELD_NAME));
             }
             if (typeOfScope.isReferenceType()) {
-                Optional<ResolvedReferenceTypeDeclaration> optionalTypeDeclaration = typeOfScope.asReferenceType().getTypeDeclaration();
-                if(optionalTypeDeclaration.isPresent()) {
-                    ResolvedReferenceTypeDeclaration typeDeclaration = optionalTypeDeclaration.get();
-                    if (typeDeclaration.isEnum()) {
-                        ResolvedEnumDeclaration enumDeclaration = (ResolvedEnumDeclaration) typeDeclaration;
-                        if (enumDeclaration.hasEnumConstant(name)) {
-                            return Optional.of(new Value(enumDeclaration.getEnumConstant(name).getType(), name));
-                        }
-                    }
-                }
-                Optional<ResolvedType> typeUsage = typeOfScope.asReferenceType().getFieldType(name);
-                return typeUsage.map(resolvedType -> new Value(resolvedType, name));
+                return solveSymbolAsValue(name, typeOfScope.asReferenceType());
+            } else if (typeOfScope.isConstraint()) {
+                return solveSymbolAsValue(name, typeOfScope.asConstraintType().getBound().asReferenceType());
             } else {
                 return Optional.empty();
             }
@@ -112,6 +104,24 @@ public class FieldAccessContext extends AbstractJavaParserContext<FieldAccessExp
                     .orElseThrow(() -> new RuntimeException("Parent context unexpectedly empty."))
                     .solveSymbolAsValue(name);
         }
+    }
+    
+    /*
+     * Try to resolve the name parameter as a field of the reference type
+     */
+    private Optional<Value> solveSymbolAsValue(String name, ResolvedReferenceType type) {
+        Optional<ResolvedReferenceTypeDeclaration> optionalTypeDeclaration = type.getTypeDeclaration();
+        if (optionalTypeDeclaration.isPresent()) {
+            ResolvedReferenceTypeDeclaration typeDeclaration = optionalTypeDeclaration.get();
+            if (typeDeclaration.isEnum()) {
+                ResolvedEnumDeclaration enumDeclaration = (ResolvedEnumDeclaration) typeDeclaration;
+                if (enumDeclaration.hasEnumConstant(name)) {
+                    return Optional.of(new Value(enumDeclaration.getEnumConstant(name).getType(), name));
+                }
+            }
+        }
+        Optional<ResolvedType> typeUsage = type.getFieldType(name);
+        return typeUsage.map(resolvedType -> new Value(resolvedType, name));
     }
 
     public SymbolReference<ResolvedValueDeclaration> solveField(String name) {

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue2823Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue2823Test.java
@@ -8,7 +8,6 @@ import com.github.javaparser.ParserConfiguration;
 import com.github.javaparser.StaticJavaParser;
 import com.github.javaparser.ast.CompilationUnit;
 import com.github.javaparser.ast.expr.FieldAccessExpr;
-import com.github.javaparser.resolution.UnsolvedSymbolException;
 import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver;
@@ -17,7 +16,7 @@ import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeS
 public class Issue2823Test extends AbstractSymbolResolutionTest {
 
     @Test
-    public void testOK() {
+    public void test() {
         final Path testRoot = adaptPath("src/test/resources/issue2823");
         TypeSolver reflectionTypeSolver = new ReflectionTypeSolver();
         JavaParserTypeSolver javaParserTypeSolver = new JavaParserTypeSolver(testRoot);
@@ -26,53 +25,20 @@ public class Issue2823Test extends AbstractSymbolResolutionTest {
                 .setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));
         StaticJavaParser.setConfiguration(configuration);
 
-        String src = 
-                "import java.util.Optional;\n" 
-                + "public class TestClass {\n" 
+        String src = "import java.util.Optional;\n"
+                + "public class TestClass {\n"
                 + "    public Long getValue() {\n"
                 + "        Optional<ClassA> classA = Optional.of(new ClassA());\n"
-                + "        return classA.map(a -> a.obj)\n" 
+                + "        return classA.map(a -> a.obj)\n"
                 + "                .map(b -> b.value)\n"
-                + "                .orElse(null);\n" 
-                + "    }\n" 
+                + "                .orElse(null);\n"
+                + "    }\n"
                 + "}";
 
         CompilationUnit cu = StaticJavaParser.parse(src);
 
-        try {
-            cu.findAll(FieldAccessExpr.class).forEach(FieldAccessExpr::resolve);
-        } catch (UnsolvedSymbolException e) {
-            // Bug to be solved
-        } 
-
-    }
-
-    @Test
-    public void testKO() {
-        final Path testRoot = adaptPath("src/test/resources/issue2823");
-        TypeSolver reflectionTypeSolver = new ReflectionTypeSolver();
-        JavaParserTypeSolver javaParserTypeSolver = new JavaParserTypeSolver(testRoot);
-        CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver(reflectionTypeSolver, javaParserTypeSolver);
-        ParserConfiguration configuration = new ParserConfiguration()
-                .setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));
-        StaticJavaParser.setConfiguration(configuration);
-
-        String src = 
-                "import java.util.Optional;\n" 
-                + "import ClassA;\n" + "import ClassB;\n"
-                + "public class TestClass {\n" 
-                + "    public Long getValue() {\n"
-                + "        Optional<ClassA> classA = Optional.of(new ClassA());\n"
-                + "        Optional<ClassB> classB = classA.map(a -> a.obj);\n" 
-                + "        return classB\n"
-                + "                .map(b -> b.value)\n" 
-                + "                .orElse(null);" 
-                + "    }\n" 
-                + "}";
-
-        CompilationUnit cu = StaticJavaParser.parse(src);
-
-        cu.findAll(FieldAccessExpr.class).forEach(FieldAccessExpr::resolve);
+        // verify there is no exception thrown when we try to resolve all field access expressions
+        cu.findAll(FieldAccessExpr.class).forEach(fd -> fd.resolve());
 
     }
 }

--- a/javaparser-symbol-solver-testing/src/test/resources/issue2823/ClassA.java
+++ b/javaparser-symbol-solver-testing/src/test/resources/issue2823/ClassA.java
@@ -1,3 +1,3 @@
 public class ClassA {
     public ClassB obj;
-}
+} 

